File: Backend/controllers/usersController.js
const User = require('../models/userModel');
const { validateUser } = require('../utils/validation');

const UsersController = {
  getAll: async (req, res) => {
    try {
      const users = await User.findAll();
      res.json({ success: true, data: users });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },
  create: async (req, res) => {
    try {
      const errors = validateUser(req.body);
      if (errors.length) return res.status(400).json({ success: false, message: errors.join(', ') });
      const id = await User.create(req.body);
      const user = await User.findById(id);
      res.status(201).json({ success: true, data: user });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },
  update: async (req, res) => {
    try {
      const affected = await User.update(req.params.id, req.body);
      if (!affected) return res.status(404).json({ success: false, message: 'User tidak ditemukan' });
      const user = await User.findById(req.params.id);
      res.json({ success: true, data: user });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },
  delete: async (req, res) => {
    try {
      if (parseInt(req.params.id) === 1) return res.status(400).json({ success: false, message: 'Tidak bisa hapus admin utama' });
      const affected = await User.delete(req.params.id);
      if (!affected) return res.status(404).json({ success: false, message: 'User tidak ditemukan' });
      res.json({ success: true, message: 'User dihapus' });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  }
};

module.exports = UsersController;

File: backend_python/controllers/users_controller.py
from flask import g, request

from models.user_model import User
from utils.helpers import fail, ok
from utils.validation import validate_user


class UsersController:
    @staticmethod
    def get_all():
        return ok({"success": True, "data": User.find_all()})

    @staticmethod
    def create():
        body = request.get_json(silent=True) or {}
        errors = validate_user(body)
        if errors:
            return fail(", ".join(errors), 400)
        user_id = User.create(body)
        return ok({"success": True, "data": User.find_by_id(user_id)}, 201)

    @staticmethod
    def update(user_id):
        body = request.get_json(silent=True) or {}
        affected = User.update(user_id, body)
        if not affected:
            return fail("User tidak ditemukan", 404)
        return ok({"success": True, "data": User.find_by_id(user_id)})

    @staticmethod
    def delete_row(user_id):
        try:
            target_id = int(user_id)
        except (TypeError, ValueError):
            return fail("User tidak ditemukan", 404)
        if target_id == 1:
            return fail("Tidak bisa hapus admin utama", 400)
        affected = User.delete(target_id)
        if not affected:
            return fail("User tidak ditemukan", 404)
        return ok({"success": True, "message": "User dihapus"})
